c/storage_mem.c

```diff
@@ -0,0 +1,286 @@
+/*
+  This program and the accompanying materials are
+  made available under the terms of the Eclipse Public License v2.0 which accompanies
+  this distribution, and is available at https://www.eclipse.org/legal/epl-v20.html
+  
+  SPDX-License-Identifier: EPL-2.0
+  
+  Copyright Contributors to the Zowe Project.
+*/
+
+#ifdef METTLE
+#error Metal C not supported
+#endif // METTLE
+
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdlib.h>
+#include <stdarg.h>
+#include <stdio.h>
+#include <limits.h>
+#include <pthread.h>
+#include "zowetypes.h"
+#include "alloc.h"
+#include "utils.h"
+#include "collections.h"
+#include "storage.h"
+
+#define STOARGE_LOCK_ERROR   (STORAGE_FIRST_CUSTOM_STATUS + 0)
+#define STOARGE_UNLOCK_ERROR (STORAGE_FIRST_CUSTOM_STATUS + 1)
+#define STORAGE_ALLOC_ERROR  (STORAGE_FIRST_CUSTOM_STATUS + 2)
+
+typedef struct MemStorage_tag {
+  hashtable *table;      // map: key -> value
+  pthread_mutex_t lock;  // lock to protect hashtable
+} MemStorage;
+
+static char *memStorageAlloc(int size) {
+  char *data = safeMalloc31(size + 4, "mem storage");
+  if (!data) {
+    return NULL;
+  }
+  *((int *)data) = size;
+  return (data + 4);
+}
+
+static void memStorageFree(char *mem) {
+  if (!mem) {
+    return;
+  }
+  char *prefix = mem - 4;
+  int allocSize = ((int *)prefix)[0];
+  safeFree31(prefix, allocSize);
+}
+
+static void freeMemStorage(void *data) {
+  memStorageFree(data);
+}
+
+#define DEFAULT_BUCKET_COUNT 257
+
+static MemStorage *makeMemStorage(MemoryStorageOptions *options) {
+  MemStorage *storage = (MemStorage*)safeMalloc(sizeof(*storage), "MemStorage");
+  if (!storage) {
+    return NULL;
+  }
+  int bucketCount = (options && options->bucketCount) > 0 ? options->bucketCount : DEFAULT_BUCKET_COUNT;
```

Comment:
Did you intend using the following?
```c
int bucketCount = options && options->bucketCount > 0 ? options->bucketCount : DEFAULT_BUCKET_COUNT;
```
Because `options && options->bucketCount` will be either 0 or 1, never -1.